### PR TITLE
Implement resources-deletion-protection including Always and Cascading

### PR DESCRIPTION
**Pull request by FillZpp**
### Ⅰ. Describe what this PR does
Implement resources-deletion-protection including Always and Cascading.

### Ⅱ. Does this pull request fix one issue?

Proposal #577 

### Ⅴ. Special notes for reviews
API definition:

- a new feature-gate named `ResourcesDeletionProtection`
- a label named `policy.kruise.io/delete-protection`, the values can be:
  - `Always`: this object will always be forbidden to be deleted, unless the label is removed
  - `Cascading`: this object will be forbidden to be deleted, if it has active resources owned

If the feature-gate has enabled, the resources below with this label will be validated for deletion operation:

| Kind                        | Group                  | Version            | Cascading judgement                                |
| --------------------------- | ---------------------- | ------------------ | ----------------------------------------------------
| `Namespace`                 | core                   | v1                 | whether there is active Pods in this namespace     |
| `CustomResourceDefinition`  | apiextensions.k8s.io   | v1beta1, v1        | whether there is existing CRs of this CRD          |
| `Deployment`                | apps                   | v1                 | whether the replicas is 0                          |
| `StatefulSet`               | apps                   | v1                 | whether the replicas is 0                          |
| `ReplicaSet`                | apps                   | v1                 | whether the replicas is 0                          |
| `CloneSet`                  | apps.kruise.io         | v1alpha1           | whether the replicas is 0                          |
| `StatefulSet`               | apps.kruise.io         | v1alpha1, v1beta1  | whether the replicas is 0                          |
| `UnitedDeployment`          | apps.kruise.io         | v1alpha1           | whether the replicas is 0                          |

### CODE REVIEW

File: apis/policy/v1alpha1/doc.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2020 The Kruise Authors.
```
Comment (hantmac) [inline apis/policy/v1alpha1/doc.go:2]: 2020 => 2021 😆

File: pkg/webhook/util/deletionprotection/deletion_protection.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	policyv1alpha1 "github.com/openkruise/kruise/apis/policy/v1alpha1"
+	"github.com/openkruise/kruise/pkg/features"
+	utilfeature "github.com/openkruise/kruise/pkg/util/feature"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func ValidateWorkloadDeletion(obj metav1.Object, replicas *int32) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ResourcesDeletionProtection) {
```
Comment (hantmac) [inline pkg/webhook/util/deletionprotection/deletion_protection.go:37]: `obj` may be nil?  There should be 
```
if !utilfeature.DefaultFeatureGate.Enabled(features.ResourcesDeletionProtection) || obj == nil {
    return nil
}
```
?

Comment (FillZpp) [reply pkg/webhook/util/deletionprotection/deletion_protection.go:37]: currently it can not be nil, but it is fine to add the protective check.

**Review by Fei-Guo — COMMENTED** [review]
LGTM

File: pkg/webhook/builtinworkloads/validating/builtin_handlers.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+func (h *WorkloadHandler) InjectDecoder(d *admission.Decoder) error {
+	h.Decoder = d
+	return nil
+}
+
+// Handle handles admission requests.
+func (h *WorkloadHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation != admissionv1beta1.Delete || req.SubResource != "" {
+		return admission.ValidationResponse(true, "")
```
Comment (furykerry) [inline pkg/webhook/builtinworkloads/validating/builtin_handlers.go:45]: replace with more meaningful func admission.Allowed("")

File: pkg/webhook/builtinworkloads/validating/builtin_handlers.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	if len(req.OldObject.Raw) == 0 {
+		klog.Warningf("Skip to validate %s %s/%s for no old object, maybe because of Kubernetes version < 1.16", req.Kind.Kind, req.Namespace, req.Name)
+		return admission.ValidationResponse(true, "")
+	}
+
```
Comment (furykerry) [inline pkg/webhook/builtinworkloads/validating/builtin_handlers.go:51]: check whether deletiontimestamp is set before continue

File: pkg/webhook/util/deletionprotection/deletion_protection.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+func ValidateNamespaceDeletion(c client.Client, namespace *v1.Namespace) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ResourcesDeletionProtection) {
+		return nil
+	}
+	switch val := namespace.Labels[policyv1alpha1.DeletionProtectionKey]; val {
+	case policyv1alpha1.DeletionProtectionTypeAlways:
+		return fmt.Errorf("forbidden by ResourcesProtectionDeletion for %s=%s", policyv1alpha1.DeletionProtectionKey, val)
+	case policyv1alpha1.DeletionProtectionTypeCascading:
+		pods := v1.PodList{}
+		if err := c.List(context.TODO(), &pods, client.InNamespace(namespace.Name)); err != nil {
```
Comment (furykerry) [inline pkg/webhook/util/deletionprotection/deletion_protection.go:61]: is it possible to disable deepcopy during list?

**Review by furykerry — CHANGES_REQUESTED** [review]

**Review by furykerry — APPROVED** [review]

**Pull request merged**